### ISSUE TITLE
 No matching model found for referenced type ... 77 more ....

### ISSUE THREAD

**Issue by fantapop** [lukeautry/tsoa#749]
I'm running into a weird one. I'm unable to build.  The error is: 
```
Error: Could not determine the keys on JobTemplateAttributes
At: ../shared/src/ipp.ts:4856:4856.
This was caused by 'keyof JobTemplateAttributes
    | 'document-copies'
    | 'document-numbers'
    | 'pages''
    at new GenerateMetadataError (/Users/chris/Development/fantapop-tsoa/src/metadataGeneration/exceptions.ts:6:5)
    at TypeResolver.resolve (/Users/chris/Development/fantapop-tsoa/src/metadataGeneration/typeResolver.ts:245:15)
    at /Users/chris/Development/fantapop-tsoa/src/metadataGeneration/typeResolver.ts:62:84
    at Array.map (<anonymous>)
    at TypeResolver.resolve (/Users/chris/Development/fantapop-tsoa/src/metadataGeneration/typeResolver.ts:61:41)
    at TypeResolver.getTypeAliasReference (/Users/chris/Development/fantapop-tsoa/src/metadataGeneration/typeResolver.ts:504:120)
    at TypeResolver.getReferenceType (/Users/chris/Development/fantapop-tsoa/src/metadataGeneration/typeResolver.ts:474:30)
    at TypeResolver.resolve (/Users/chris/Development/fantapop-tsoa/src/metadataGeneration/typeResolver.ts:307:32)
    at TypeResolver.resolve (/Users/chris/Development/fantapop-tsoa/src/metadataGeneration/typeResolver.ts:55:133)
    at TypeResolver.propertyFromSignature (/Users/chris/Development/fantapop-tsoa/src/metadataGeneration/typeResolver.ts:827:113)
```
I've got a local copy of tsoa running so I was trying to debug a bit.  That area of the code was catching an error but not logging it.  When I log it out I get:

```
error was:  Error: No matching model found for referenced type ... 77 more .... If ... 77 more ... comes from a dependency, please create an interface in your own code that has the same structure. Tsoa can not utilize interfaces from external dependencies. Read more
 at https://github.com/lukeautry/tsoa/blob/master/docs/ExternalInterfacesExplanation.MD
```

My best guess here is that there are too many classes and typescript isn't giving all the info so at some point, it gives back "77 more" expecting to display that in a gui. 

Here is some more info:

getModelTypeDeclaration is being called with this type: 
```
 IdentifierObject {
    pos: -1,
    end: -1,
    flags: 8,
    modifierFlagsCache: 0,
    transformFlags: 0,
    parent: undefined,
    kind: 75,
    escapedText: '... 77 more ...',
    originalKeywordKind: undefined,
    autoGenerateFlags: 0,
    autoGenerateId: 0
  }
```

- **I'm submitting a ...**

  - [x] bug report
  - [ ] feature request
  - [ ] support request

- I confirm that I
  - [x] used the [search](https://github.com/lukeautry/tsoa/search?type=Issues) to make sure that a similar issue hasn't already been submit

## Possible Solution

I'm not sure what else to try here. I'm open to debugging some more with some instruction.

## Context (Environment)

Version of the library: 3.2.1
Version of NodeJS: I'm running tsoa using ts-node 

- Confirm you were using yarn not npm: [x]

**Comment by jookshub** [comment]
I've encountered a similar kind of issue and used a `console.log('parentNode:', parentNode)` in `typeResolver.js`:

```javascript
parentNode: NodeObject {
  pos: -1,
  end: -1,
  flags: 8,
  modifierFlagsCache: 0,
  transformFlags: 0,
  parent: undefined,
  kind: 158,
  modifiers: undefined,
  name: IdentifierObject {
    pos: -1,
    end: -1,
    flags: 8,
    modifierFlagsCache: 0,
    transformFlags: 0,
    parent: undefined,
    kind: 75,
    escapedText: '... 4 more ...',
    originalKeywordKind: undefined,
    autoGenerateFlags: 0,
    autoGenerateId: 0
  },
  questionToken: undefined,
  type: undefined,
  initializer: undefined
}
Generate swagger error.
 TypeError: Cannot read property 'kind' of undefined
    at TypeResolver.getPrimitiveType (/Users/jokroj/__Cellular__/bitburger-one-cds/node_modules/tsoa/dist/metadataGeneration/typeResolver.js:349:72)
    at TypeResolver.resolve (/Users/jokroj/__Cellular__/bitburger-one-cds/node_modules/tsoa/dist/metadataGeneration/typeResolver.js:92:34)
    at /Users/jokroj/__Cellular__/bitburger-one-cds/node_modules/tsoa/dist/metadataGeneration/typeResolver.js:147:118
    at Array.reduce (<anonymous>)
    at TypeResolver.resolve (/Users/jokroj/__Cellular__/bitburger-one-cds/node_modules/tsoa/dist/metadataGeneration/typeResolver.js:146:18)
    at /Users/jokroj/__Cellular__/bitburger-one-cds/node_modules/tsoa/dist/metadataGeneration/typeResolver.js:147:118
    at Array.reduce (<anonymous>)
    at TypeResolver.resolve (/Users/jokroj/__Cellular__/bitburger-one-cds/node_modules/tsoa/dist/metadataGeneration/typeResolver.js:146:18)
    at TypeResolver.resolve (/Users/jokroj/__Cellular__/bitburger-one-cds/node_modules/tsoa/dist/metadataGeneration/typeResolver.js:308:118)
    at MethodGenerator.Generate (/Users/jokroj/__Cellular__/bitburger-one-cds/node_modules/tsoa/dist/metadataGeneration/methodGenerator.js:52:76)
```

If I decrease the size of an object I am returning in one of my functions this error is omitted. And by size I mean that using a different length of a field name e.g. *demo* instead of *longdemo*.

I'll try to reduce this down into a neat example, but since this may take a while, I'll leave this comment to support further investigation of this bug.

**Comment by fantapop** [comment]
not stale.  Maybe I will have time to look more into this. I see these in the VS Code type hints as well for large types.  I found this: https://stackoverflow.com/questions/53113031/how-to-see-a-fully-expanded-typescript-type-without-n-more-and  maybe setting this will help.

**Comment by fantapop** [comment]
Also this: https://github.com/Microsoft/TypeScript/issues/26238

**Comment by WoH** [comment]
I'd hope there is a way to access the properties without truncation regardless of the project config using the compiler/tc api

**Comment by fantapop** [comment]
It seems like the action bot decided to not respect commenting in order to remove the stale label

Anyways, I was able to make some progress. I got passed this error by doing something similar to what is described here: https://github.com/microsoft/TypeScript/issues/26238#issuecomment-672086446

In the built typescript files, there was being many copies of the mentioned line: `ts.defaultMaximumTruncationLength = 160`

I modified this file: node_modules/typescript/lib/typescript.js. My type was apparently very large. I had to bump up the `ts.defaultMaximumTruncationLength` to 2000 to finally make my build succeed.

I also found mention of a method to set this value here:
https://github.com/microsoft/TypeScript/issues/29564

by doing like `(ts as any).defaultMaximumTruncationLength = 2000;`

however, I've been unable to get this to have an affect by setting the value from inside TSOA.

**Comment by jookshub** [comment]
Thanks to @fantapop and https://github.com/microsoft/TypeScript/issues/26238#issuecomment-672086446 I was able to solve this with ```ts.NodeBuilderFlags.NoTruncation```

I've created a fork and a patch-branch here:
https://github.com/lukeautry/tsoa/compare/master...cellular:feature/patch_truncation_issue_749

However, I am lost in providing a test for that. All tests run trough - with and without the patch.

I've tried to edit the ```ModelService``` and the ```TestClassModel```, but failed to edit them in a way, that would result in the error this issue is about.

So a patch is available, a test ist missing.
